src/Client/HedgedConnectionsFactory.cpp

```diff
@@ -0,0 +1,360 @@
+{
+    /// Check if there is no free replica.
+    if (entries_count + replicas_in_process_count + failed_pools_count >= shuffled_pools.size())
+        return -1;
+
+    /// Check if it's the first time.
+    if (last_used_index == -1)
+    {
+        last_used_index = 0;
+        return 0;
+    }
+
+    bool finish = false;
+    int next_index = last_used_index;
+    while (!finish)
+    {
+        next_index = (next_index + 1) % shuffled_pools.size();
+
+        /// Check if we can try this replica.
+        if (!replicas[next_index].connection_establisher.isInProcess()
+            && !replicas[next_index].connection_establisher.isFinished()
+            && (max_tries == 0 || shuffled_pools[next_index].error_count < max_tries))
+            finish = true;
+
+        /// If we made a complete round, there is no replica to connect.
+        else if (next_index == last_used_index)
+            return -1;
+    }
+
+    last_used_index = next_index;
+    return next_index;
+}
+
+int HedgedConnectionsFactory::startEstablishingNewConnection(Connection *& connection_out)
+{
+    int index;
+    do
+    {
+        index = getNextIndex();
+        if (index == -1)
+            return -1;
+
+        ReplicaStatus & replica = replicas[index];
+        ++replicas_in_process_count;
+        replica.connection_establisher.resume();
+
+        processConnectionEstablisherStage(index);
+
+        if (replica.connection_establisher.isInProcess())
+        {
+            epoll.add(replica.connection_establisher.getFileDescriptor());
+            fd_to_replica_index[replica.connection_establisher.getFileDescriptor()] = index;
+
+            /// Add timeout for changing replica.
+            replica.change_replica_timeout.setRelative(timeouts.hedged_connection_timeout);
+            epoll.add(replica.change_replica_timeout.getDescriptor());
+            timeout_fd_to_replica_index[replica.change_replica_timeout.getDescriptor()] = index;
+        }
+    }
+    while (!replicas[index].connection_establisher.isInProcess() && !replicas[index].is_ready);
+
+    if (replicas[index].is_ready)
+        connection_out = replicas[index].connection_establisher.getConnection();
+
+    return index;
+}
+
+void HedgedConnectionsFactory::processConnectionEstablisherStage(int index, bool remove_from_epoll)
+{
+    ReplicaStatus & replica = replicas[index];
+
+    if (replica.connection_establisher.isFinished())
+    {
+        --replicas_in_process_count;
+        ++entries_count;
+
+        if (remove_from_epoll)
+            removeReplicaFromEpoll(index);
+
+        if (replica.connection_establisher.getResult().is_usable)
+        {
+            ++usable_count;
+            if (replica.connection_establisher.getResult().is_up_to_date)
+                replica.is_ready = true;
+
+            return;
+        }
+
+        /// If replica is not usable, we need to save fail message.
+        if (!replica.connection_establisher.getFailMessage().empty())
```

Comment:
I was a little bit unusual for me that we can get `isFinished` result, but **not** `is_usable` connection.
It seem to be possible in case if replica does not have table (from code in `connection_establisher`).
However, wouldn't it be more logical to return `isFailed` in this case? As soon as we cannot process this case anyway ...